src/GafferImage/FlatToDeep.cpp

```diff
@@ -0,0 +1,352 @@
+
+using namespace std;
+using namespace Imath;
+using namespace IECore;
+using namespace GafferImage;
+using namespace Gaffer;
+
+GAFFER_GRAPHCOMPONENT_DEFINE_TYPE( FlatToDeep );
+
+size_t FlatToDeep::g_firstPlugIndex = 0;
+
+FlatToDeep::FlatToDeep( const std::string &name )
+	:	ImageProcessor( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+	addChild( new FloatPlug( "depth" ) );
+	addChild( new StringPlug( "sourceZChannel" ) );
+
+	addChild( new FloatPlug( "thickness", Plug::In, 0.0f, 0.0f ) );
+	addChild( new StringPlug( "sourceZBackChannel" ) );
+
+
+	// Pass-through the things we don't want to modify.
+	outPlug()->formatPlug()->setInput( inPlug()->formatPlug() );
+	outPlug()->dataWindowPlug()->setInput( inPlug()->dataWindowPlug() );
+	outPlug()->metadataPlug()->setInput( inPlug()->metadataPlug() );
+}
+
+FlatToDeep::~FlatToDeep()
+{
+}
+
+FloatPlug *FlatToDeep::depthPlug()
+{
+	return getChild<FloatPlug>( g_firstPlugIndex );
+}
+
+const FloatPlug *FlatToDeep::depthPlug() const
+{
+	return getChild<FloatPlug>( g_firstPlugIndex );
+}
+
+StringPlug *FlatToDeep::sourceZChannelPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex+1 );
+}
+
+const StringPlug *FlatToDeep::sourceZChannelPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex+1 );
+}
+
+FloatPlug *FlatToDeep::thicknessPlug()
+{
+	return getChild<FloatPlug>( g_firstPlugIndex+2 );
+}
+
+const FloatPlug *FlatToDeep::thicknessPlug() const
+{
+	return getChild<FloatPlug>( g_firstPlugIndex+2 );
+}
+
+StringPlug *FlatToDeep::sourceZBackChannelPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex+3 );
+}
+
+const StringPlug *FlatToDeep::sourceZBackChannelPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex+3 );
+}
+
+void FlatToDeep::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	ImageProcessor::affects( input, outputs );
+
+	if( input == inPlug()->channelNamesPlug() )
+	{
+		outputs.push_back( outPlug()->channelNamesPlug() );
+	}
+	else if( input == inPlug()->channelDataPlug() )
+	{
+		outputs.push_back( outPlug()->channelDataPlug() );
+	}
+	else if( input == sourceZChannelPlug() ||
+		input == depthPlug() ||
+		input == sourceZBackChannelPlug() ||
+		input == thicknessPlug()
+	)
+	{
+		outputs.push_back( outPlug()->channelDataPlug() );
+	}
+}
+
+void FlatToDeep::hashChannelNames( const GafferImage::ImagePlug *output, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	ImageProcessor::hashChannelNames( output, context, h );
+	inPlug()->channelNamesPlug()->hash( h );
+}
+
+IECore::ConstStringVectorDataPtr FlatToDeep::computeChannelNames( const Gaffer::Context *context, const ImagePlug *parent ) const
+{
+	StringVectorDataPtr resultData = inPlug()->channelNamesPlug()->getValue()->copy();
+	vector<string> &result = resultData->writable();
+
+	if( find( result.begin(), result.end(), "Z" ) == result.end() )
+	{
+		result.push_back( "Z" );
+	}
+
+	if( find( result.begin(), result.end(), "ZBack" ) == result.end() )
```

Comment:
I had a comment asking this myself earlier, but ended up deciding to delete the comment and leave it as is.    You always want me to keep user features as minimal as possible.

If you think it's appropriate, just tell me what you want the UI to be, it would be easy to code.  I really don't much like how FlatToDeep handles the depth and thickness / ZBack plug names, but haven't come up with anything better.

Reply:
> You always want me to keep user features as minimal as possible.

Yep, "as simple as possible but no simpler". I wonder if we're going too simple here, or rather, if FlatToDeep is just formulated a bit oddly anyway.

> I really don't much like how FlatToDeep handles the depth and thickness / ZBack plug names, but haven't come up with anything better.

Yeah, it feels odd to me too. For one, it seems like it's maybe a bit of an anti-pattern to have `sourceZChannel` disable the plug _above_ it. Typically we present things such that if you worked through the UI from top to bottom, you wouldn't be doing anything irrelevant.

How about this?

zMode : Dropdown with a value of either "Constant" or "Channel".
depth : The constant used when in "Constant" mode. Hidden when in "Channel" mode.
zChannel : The channel used when in "Channel" mode. Hidden when in "Constant" mode.

zBackMode : Dropdown with a value of either "None", "Thickness" or "Channel", defaulted to "None"
thickness : The constant used when in "Thickness" mode. Hidden in the other modes.
zBackChannel : The channel used when in "Channel" mode. Hidden in the other modes.

There are two more plugs there, but we'd still only be showing four at any point in time. We get to omit ZBack if we want, and I think the UI becomes a bit more intuitive. Any other thoughts?